bot/exts/moderation/watchchannels/_watchchannel.py

```diff
@@ -366,8 +369,6 @@ async def prepare_watched_users_data(
     def _remove_user(self, user_id: int) -> None:
         """Removes a user from a watch channel."""
         self.watched_users.pop(user_id, None)
-        self.message_queue.pop(user_id, None)
-        self.consumption_queue.pop(user_id, None)
```

Comment:
The watched_users pop here remains, but can't the same issue occur if you're trying to remove a user and list the users at the same time? Maybe it's worth converting the iterator in `prepare_watched_users_data` into a list before looping over it.

For the other two dicts, does this definitely resolve the issue? Looking at `on_message`, you could theoretically have multiple `consume_messages` tasks. If one task clears a queue while another is iterating over it, won't it create the same bug?

Reply:
> The watched_users pop here remains, but can't the same issue occur if you're trying to remove a user and list the users at the same time? Maybe it's worth converting the iterator in prepare_watched_users_data into a list before looping over it.

I guess so, I'll do that.

> For the other two dicts, does this definitely resolve the issue? Looking at on_message, you could theoretically have multiple consume_messages tasks. If one task clears a queue while another is iterating over it, won't it create the same bug?

As soon as one task is created, the `consuming_messages` check will stop any more from being created while it's running: https://github.com/python-discord/bot/blob/bfd9e446d4d242dd8948057ed3e29faaa699267f/bot/exts/moderation/watchchannels/_watchchannel.py#L76-L90

Reply:
https://github.com/python-discord/bot/pull/3017/commits/25acba42f18b7f5ef813dd7ed5d62cead9f3940b